### PR TITLE
Fix overlong function signature

### PR DESCRIPTION
**Pull request by dawirstejeck**
Fix off-by-one error in rewrite_fn_base.  #1049 

Also remove a line break.

### CODE REVIEW

**Comment by pepyakin** [comment]
Hi, consider following test case:

```rust
impl Handle {
    pub fn merge(a: Abcd) -> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Internal>, marker::Edge> {
    }
}
```

I suspect, that there is more than one-by-off issue. We can be under `max_width` but still can exceed maximum length after adding ```" {"```.

**Comment by dawirstejeck** [comment]
You are right. I try to fix it tomorrow.

File: src/items.rs [hunk 1]
```diff
@@ -1359,8 +1359,9 @@ fn rewrite_fn_base(context: &RewriteContext,
             _ => {
                 // If we've already gone multi-line, or the return type would push
                 // over the max width, then put the return type on a new line.
-                result.contains("\n") || multi_line_ret_str ||
-                result.len() + indent.width() + ret_str_len > context.config.max_width
+                let overlong_ret = result.len() + indent.width() + ret_str_len >
```
Comment (nrc) [inline src/items.rs]: Could you add a comment explaining why we need the `-1`

File: tests/source/issue-1049.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Test issue-1049
```
Comment (nrc) [inline tests/source/issue-1049.rs]: Could you explain what is being tested here please?

**Comment by dawirstejeck** [comment]
Will continue to work on this and the other issue next weekend. No spare time until then.

**Comment by dawirstejeck** [comment]
Nightly seems to have problems but i think it is not because of my changes.

**Comment by nrc** [comment]
Yeah, the nightly issue is probably a bug in Rust.

**Comment by nrc** [comment]
And in fact that should be fixed now. I restarted the tests here.

File: src/items.rs [hunk 3]
```diff
@@ -1357,10 +1357,21 @@ fn rewrite_fn_base(context: &RewriteContext,
...
-                result.len() + indent.width() + ret_str_len > context.config.max_width
+                // If we've already gone multi-line, or the return type would push over the max
+                // width, then put the return type on a new line. With the +1 for the signature
+                // length an additional space between the closing brace of the argument and the
+                // arrow '->' is considered.
+                let mut sig_length = result.len() + indent.width() + ret_str_len + 1;
+
+                // If there is no where clause, take into account the space after the return type
+                // and the brace.
+                if where_clause.predicates.is_empty() {
```
Comment (nrc) [inline src/items.rs]: Do we need to check if the return type exists? I.e., not do this if there is no return type?

Also, nit, brace = `{}`, I think you mean parenthesis in this comment and above.

Comment (dawirstejeck) [inline src/items.rs]: Yes, you are right. No return type means no arrow means no additional space.

Comment (juicejitsu) [inline src/items.rs]: Isn't this checked on ln 1354 already? ` if !ret_str.is_empty()`

**Comment by juicejitsu** [comment]
@dawirstejeck whats left to do here? lgtm

File: src/items.rs (hunk 3 above)
Comment (dawirstejeck) [inline src/items.rs]: Oh, I overlooked that... :sweat_smile: but to be sure I will add a test for that case later.

**Comment by dawirstejeck** [comment]
@juicejitsu you are probably right. But thank you for having a look at it!

PS: I'm sorry for being so slow. I love to participate but I only got 1 or 2 hours per week :snail:.

**Comment by nrc** [comment]
@juicejitsu, @dawirstejeck: does this PR cover everything that #1108 does? Is there a reason to land one or the other?

**Comment by juicejitsu** [comment]
@nrc yeah this covers the bug in that PR as well, although there are some extra test case that are not in this one.

**Comment by nrc** [comment]
Thank you!

**Pull request merged**